## Constructing Feedfoward Neural Networks (FNNs) with TensorFlow

Here, we construct a FNN with three hidden layers.

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models

# Extended model with two hidden layers
model = models.Sequential ( [
layers.Input(shape = (5,)), # Explicit input layer avoids errors
layers.Dense ( 128 , activation = 'relu') ,
# First hidden layer
layers.Dense ( 64 , activation = 'relu') ,
# Second hidden layer
layers.Dense ( 32 , activation = 'relu') ,
# Third hidden layer
layers.Dense ( 10 , activation = 'softmax')
# Output layer
] )

Now, we run `model.summary()`:

In [2]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 128)            │           768 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │           330 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 11,434 (44.66 KB)

 Trainable params: 11,434 (44.66 KB)

 Non-trainable params: 0 (0.00 B)

The total number of parameters in the model is calculated as follows (weights + biases!):
- First layer: 5*128 + 128 = 768
- Second layer: 128*64 + 64 = 8256
- Third layer: 64*32 + 32 = 2080
- Output layer: 32*10 + 10 = 330
- Total: 11,434

## PyTorch Demonstration

In [6]:
import torch
import torch.nn as nn
# Define a simple feed-forward neural network
class SimpleNN (nn.Module) :
    def __init__ (self) :
        super (SimpleNN, self ).__init__()
        self.fc1 = nn.Linear(5, 128) # Input layer
        self.fc2 = nn.Linear(128 , 64) # Hidden layer
        self.fc3 = nn.Linear(64, 10) # Output layer
    def forward (self , x):
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return x
# Instantiate the model
model = SimpleNN()

# Print the model architecture
print(model)

# Calculate the number of parameters
total_params = sum (p.numel() for p in model.parameters())
trainable_params = sum (p.numel() for p in model.parameters() if p.requires_grad)

print (f'Total parameters : {total_params}')
print (f'Trainable parameters : {trainable_params}')

SimpleNN(
  (fc1): Linear(in_features=5, out_features=128, bias=True)
  (fc2): Linear(in_features=128, out_features=64, bias=True)
  (fc3): Linear(in_features=64, out_features=10, bias=True)
)
Total parameters : 9674
Trainable parameters : 9674
